freud/diffraction.pyx

```diff
@@ -0,0 +1,313 @@
+# Copyright (c) 2010-2020 The Regents of the University of Michigan
+# This file is from the freud project, released under the BSD 3-Clause License.
+
+R"""
+The :class:`freud.diffraction` module provides functions for computing the
+diffraction pattern of particles in systems with long range order.
+
+.. rubric:: Stability
+
+:mod:`freud.diffraction` is **unstable**. When upgrading from version 2.x to
+2.y (y > x), existing freud scripts may need to be updated. The API will be
+finalized in a future release.
+"""
+
+import freud.locality
+import logging
+import numpy as np
+import scipy.ndimage
+import rowan
+
+from freud.util cimport _Compute
+cimport numpy as np
+
+
+logger = logging.getLogger(__name__)
+
+
+cdef class DiffractionPattern(_Compute):
+    R"""Computes a 2D diffraction pattern.
+
+    The diffraction image represents the scattering of incident radiation,
+    and is useful for identifying translational and/or rotational symmetry
+    present in the system. This class computes the static `structure factor
+    <https://en.wikipedia.org/wiki/Structure_factor>`_ :math:`S(\vec{q})` for
+    a plane of wavevectors :math:`\vec{q}` orthogonal to a view axis. The
+    view orientation :math:`(1, 0, 0, 0)` defaults to looking down the
+    :math:`z` axis (at the :math:`xy` plane). The points in the system are
+    converted to fractional coordinates, then binned into a grid whose
+    resolution is given by ``grid_size``. A higher ``grid_size`` will lead to
+    a higher resolution. The points are convolved with a Gaussian of width
+    :math:`\sigma`, given by ``peak_width``. This convolution is performed
+    as a multiplication in Fourier space. The computed diffraction pattern
+    is returned as a square array of shape ``(output_size, output_size)``.
+
+    Args:
+        grid_size (unsigned int):
+            Resolution of the diffraction grid (Default value = 512).
+        output_size (unsigned int):
+            Size of the output diffraction image (Default value = 512).
+        zoom (float):
+            Scaling factor for incident wavevectors (Default value = 4).
+        peak_width (float):
+            Width of Gaussian convolved with points, in system length units
+            (Default value = 1).
+    """
+    cdef int grid_size
+    cdef int output_size
+    cdef double zoom
+    cdef double peak_width
+    cdef double[:] _k_values_orig
+    cdef double[:, :, :] _k_vectors_orig
+    cdef double[:] _k_values
+    cdef double[:, :, :] _k_vectors
+    cdef double[:, :] _diffraction
+
+    def __init__(self, grid_size=512, output_size=512, zoom=4, peak_width=1):
```

Comment:
`zoom` and `peak_width` can be moved to arguments of the `compute` method to enable faster repeated calculations of the same `grid_size`, but different `peak_width` and `zoom`. `output_size` could be moved as well given the state the code is in now, but I suspect that there are small problems with the current state of the code so it shouldn't be that way when we merge this PR.

Reply:
We cannot move `output_size` from the constructor to `compute` because the cached `_k_vectors_orig` and `_k_values_orig` rely on `output_size`. However, I think this is a good idea to move `zoom` and `peak_width` so they can be changed by the user interactively (which would also be useful if this feature were being used in a graphical interface).